api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto

```diff
@@ -50,4 +51,37 @@ message RedisProxy {
 
   // Network settings for the connection pool to the upstream cluster.
   ConnPoolSettings settings = 3 [(validate.rules).message.required = true];
+
+  message RouteMatch {
+    // If specified, the route is a prefix rule that must match the beginning of the keys. The proxy
+    // will always favor the longest matches.
+    string prefix = 1 [(validate.rules).string.min_bytes = 1];
+
+    // Indicates that prefix/path matching should be case insensitive. The default is true.
+    google.protobuf.BoolValue case_sensitive = 2;
+  }
+
+  message RouteAction {
+    // Name of cluster from cluster manager. See the :ref:`configuration section
+    // <arch_overview_redis_configuration>` of the architecture overview for recommendations on
+    // configuring the backing cluster.
+    string cluster = 1 [(validate.rules).string.min_bytes = 1];
+
+    // Network settings for the connection pool to the upstream cluster.
+    ConnPoolSettings settings = 2;
+  }
+
+  message Route {
+    // Route matching parameters.
+    RouteMatch match = 1 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
+
+    // Route request to some upstream cluster.
+    RouteAction route = 2 [(validate.rules).message.required = true, (gogoproto.nullable) = false];
+  }
+
+  message RouteConfig {
+    repeated Route routes = 1 [(gogoproto.nullable) = false];
+  }
+
+  RouteConfig route_config = 4 [(gogoproto.nullable) = false];
```

Comment:
Can you add documentation/comments on how this field interacts with the top level cluster name and conn pool settings? Is that a required catch all if no route matches? If not, should we consider deprecating that and requiring people to specify a single route entry that matches everything?

Reply:
The way I currently implemented it was by using the original `cluster + settings` tuple to create a catch all route. This way it doesn't break backward compatibility, but it makes it a bit harder to understand what it's used for. 

I was also thinking that if no routes are defined, the default `cluster + settings` becomes mandatory. But if some routes are defined, they become optional since I think there's valid scenarios where you wouldn't want any catch all instance and fail loudly.

Any thoughts?

Reply:
Personally, I think it would be most clear to just deprecate the top level cluster/settings, and then make it an error if they are set when any routes are set (basically we should require people to be explicit if they are using routes). Then in the future we can just delete the deprecated option and then people just have to specify at least one route which would be a catch-all. WDYT?

Reply:
I gave this implementation a shot in https://github.com/envoyproxy/envoy/pull/5658/commits/f822dd83c1144c61b23c85fc8e8557356c7fd983, let me know what you think!